substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/DwarfARangesSectionImpl.java

```diff
@@ -231,76 +193,60 @@ public void writeContent(DebugContext context) {
-            pos = writeShort(DwarfDebugInfo.DW_VERSION_2, buffer, pos);
-            pos = writeInt(cuIndex, buffer, pos);
-            /* Address size is always 8. */
-            pos = writeByte((byte) 8, buffer, pos);
-            /* Segment size is always 0. */
-            pos = writeByte((byte) 0, buffer, pos);
-            assert (pos - lastpos) == DW_AR_HEADER_SIZE;
-            /*
-             * Align to 2 * address size.
-             */
-            for (int i = 0; i < DW_AR_HEADER_PAD_SIZE; i++) {
-                pos = writeByte((byte) 0, buffer, pos);
-            }
-            log(context, "  [0x%08x] Address          Length           Name", pos);
-            for (PrimaryEntry classPrimaryEntry : classPrimaryEntries) {
-                Range primary = classPrimaryEntry.getPrimary();
-                /*
-                 * Emit only linkage stubs.
-                 */
-                if (primary.isDeoptTarget()) {
-                    log(context, "  [0x%08x] %016x %016x %s", pos, debugTextBase + primary.getLo(), primary.getHi() - primary.getLo(), primary.getFullMethodNameWithParams());
-                    pos = writeRelocatableCodeOffset(primary.getLo(), buffer, pos);
-                    pos = writeLong(primary.getHi() - primary.getLo(), buffer, pos);
-                }
-            }
+            log(context, "  [0x%08x] %s CU (deopt) %d ", pos, classEntry.getFileName(), cuIndex);
+            pos = writeHeader(context, cuIndex, buffer, pos);
+            pos = writeARanges(context, classEntry.deoptPrimaryEntries(), buffer, pos);
+            // write two terminating zeroes
             pos = writeLong(0, buffer, pos);
             pos = writeLong(0, buffer, pos);
+            // backpatch the length field
+            patchLength(lengthPos, buffer, pos);
         }
         assert pos == size;
     }
 
-    public static int compare(int pc1, int pc2) {
-        if (pc1 < pc2) {
-            return -1;
-        } else if (pc1 > pc2) {
-            return 1;
-        } else {
-            return 0;
+    private int writeHeader(DebugContext context, int cuIndex, byte[] buffer, int p) {
+        int pos = p;
+        // write dummy length for now
+        pos = writeInt(0, buffer, pos);
+        /* DWARF version is always 2. */
+        pos = writeShort(DwarfDebugInfo.DW_VERSION_2, buffer, pos);
+        pos = writeInt(cuIndex, buffer, pos);
+        /* Address size is always 8. */
+        pos = writeByte((byte) 8, buffer, pos);
+        /* Segment size is always 0. */
+        pos = writeByte((byte) 0, buffer, pos);
+        assert (pos - p) == DW_AR_HEADER_SIZE;
+        /*
+         * Align to 2 * address size.
+         */
+        for (int i = 0; i < DW_AR_HEADER_PAD_SIZE; i++) {
+            pos = writeByte((byte) 0, buffer, pos);
         }
+        return pos;
+    }
+
+    int writeARanges(DebugContext context, Stream<PrimaryEntry> entries, byte[] buffer, int p) {
+        return entries.reduce(p,
```

Comment:
Using operations with side effects with stream methods like `reduce` is not recommend. It's better to use `forEach` (or even `forEachOrdered`) instead. Final `pos` is just `p + buffer_length_delta` right?

Reply:
It's not just about knowing what the final position is. This reduction operation is essentially using a stream to implement an indexed for loop with index var `p1` and loop step computed by the write calls in the loop body. The index is used as a cursor into byte array `buffer` to identify where to start writing the range info for each successive method. The cursor is consumed and updated by the two write operations. The final lambda is used to pass the updated cursor value on to defined `p1` for the next execution of the loop body. The final cursor value pops out at the end /defining/ the final position.

Now, in this case each iteration appends the same fixed amount to the buffer. So, you are right to observe that we can compute in advance what the final cursor position would be. However, I don't think that means we can simply switch to a `forEach`. We still need to inject a cursor start position into each invocation of the lambda in order to write each successive method's range data at the appropriate offset.

A better alternative, one that would be usable with stream `forEach`, would be to wrap the byte array in an object which also embeds a cursor updated by each write and pass that to the lambda as a final object. We could define our own wrapper or rely on an existing type like ByteBuffer or ByteArrayOutputStream. That would avoid the need for a reduction.

However, it would be much better if we used a common type across the code base, one which parameterized all the `writeXXX` methods. This type should also really provide a base API that is DWARF aware, implementing the various 'putXXX' methods that the `writeXXX` methods depend on (since these are the ones that would actually need to update the cursor). That would work fine with a `forEach` model so long as all uses were sequential.

This really should have been done from the outset. Unfortunately when I started it was not clear what was needed so I simply started/stuck with peeking and poking bytes into a byte array using an externally maintained offset. I was relying on lists and iterators at that point, not streams, so using a byte buffer and explicit buffer index was not such an issue. It certainly would simplify and clarify the code if we switched to using some form of buffer/stream which maintained its own cursor position. However, I think it would be better to do this as a global clean-up rather than as a local or global change mixed in with this patch.

Reply:
Ok, I had a rethink and came up with a relatively simple solution that 1) removes the reductions and 2) can either be extended later to deal with all cases or can be superseded by use of a ByteBuffer/ByteArrayOutputStream class.

I defined a class called `Cursor` nested in `DwarfSectionImpl` that wraps an `int` field called `pos`, with the obvious `get`, `set` and `add` methods. This allows it to be used in all the `DwarfXXXSection` classes when they are writing data to `byte[] content`.

A `Cursor` can be created and set to the current `pos` (or default to 0)  at any point where we want to iterate over a stream of class entries, compiled method entries etc. The cursor can be passed into the `forEach` lambda bodies and `get`, `set` or `add` can be used to read and/or update the current position.

This allows all existing reduce operations to be simplified to a `forEach` -- so long as it executes sequentially. We could extend this so that all `writeXXX` and `putXXX` operations take a `Cursor pos` argument in place of the current `int pos` argument. Alternatively, we can switch to using some other type to encapsulate both the `byte[] buffer` and `int pos`.

I'll push a patch soon.